File: YukariLauncher/src/main/java/net/kdt/pojavlaunch/utils/FileUtils.java
package net.kdt.pojavlaunch.utils;

import java.io.File;
import java.io.IOException;

public class FileUtils {
    /**
     * 文字列パスで指定されたファイルが存在するか確認する。
     * @param filePath 確認するファイルパス
     * @return ファイルが存在する場合はtrue（File.exists()と同じ）
     */
    public static boolean exists(String filePath){
        return new File(filePath).exists();
    }

    /**
     * パスまたはURL文字列からファイル名を取得する。
     * @param pathOrUrl ファイルのパスまたはURL
     * @return ファイル名。スラッシュがない場合はnull
     */
    public static String getFileName(String pathOrUrl) {
        int lastSlashIndex = pathOrUrl.lastIndexOf('/');
        if(lastSlashIndex == -1) return null;
        return pathOrUrl.substring(lastSlashIndex);
    }

    /**
     * パスまたはURL文字列から拡張子（最後のドット以降）を削除する。
     * @param pathOrUrl ファイルのパスまたはURL
     * @return 拡張子が削除された文字列
     */
    public static String removeExtension(String pathOrUrl) {
        int lastDotIndex = pathOrUrl.lastIndexOf('.');
        if(lastDotIndex == -1) return pathOrUrl;
        return pathOrUrl.substring(0, lastDotIndex);
    }

    /**
     * ディレクトリが存在し、ディレクトリであり、書き込み可能であることを確認する。
     * @param targetFile 確認するディレクトリ
     * @return 確認に成功した場合はtrue
     */
    public static boolean ensureDirectorySilently(File targetFile) {
        if(targetFile.isFile()) return false;
        if(targetFile.exists()) return targetFile.canWrite();
        else return targetFile.mkdirs();

    }

    /**
     * ファイルの親ディレクトリが存在し、書き込み可能であることを確認する。
     * @param targetFile 親ディレクトリを確認するファイル
     * @return 確認に成功した場合はtrue
     */
    public static boolean ensureParentDirectorySilently(File targetFile) {
        File parentFile = targetFile.getParentFile();
        if(parentFile == null) return false;
        return ensureDirectorySilently(parentFile);
    }

    /**
     * ensureDirectorySilently()と同じだが、失敗時にIOExceptionをスローする。
     * @param targetFile 確認するディレクトリ
     * @throws IOException 確認に失敗した場合
     */
    public static void ensureDirectory(File targetFile) throws IOException{
        if(targetFile.isFile()) throw new IOException("Target directory is a file");
        if(targetFile.exists()) {
            if(!targetFile.canWrite()) throw new IOException("Target directory is not writable");
        }else if(!targetFile.mkdirs()) throw new IOException("Unable to create target directory");
    }

    /**
     * ensureParentDirectorySilently()と同じだが、失敗時にIOExceptionをスローする。
     * @param targetFile 親ディレクトリを確認するファイル
     * @throws IOException 確認に失敗した場合
     */
    public static void ensureParentDirectory(File targetFile) throws IOException{
        File parentFile = targetFile.getParentFile();
        if(parentFile == null) throw new IOException("targetFile does not have a parent");
        ensureDirectory(parentFile);
    }
}

File: YukariLauncher/src/main/java/com/arata/yukarilauncher/utils/file/FileUtils.kt
package com.arata.yukarilauncher.utils.file

import java.io.File
import java.io.IOException

/** ファイル・ディレクトリ操作のユーティリティを提供するオブジェクト。 */
object FileUtils {

    /** 指定されたパスのファイルまたはディレクトリが存在するか確認する。 @param filePath 確認するパス @return 存在する場合は true */
    fun exists(filePath: String): Boolean {
        return File(filePath).exists()
    }

    /** パスまたはURLからファイル名を取得する。 @param pathOrUrl パスまたはURL @return ファイル名、取得不可の場合は null */
    fun getFileName(pathOrUrl: String): String? {
        val lastSlashIndex = pathOrUrl.lastIndexOf('/')
        if (lastSlashIndex == -1) return null
        return pathOrUrl.substring(lastSlashIndex)
    }

    /** パスまたはURLから拡張子を取り除く。 @param pathOrUrl パスまたはURL @return 拡張子除去後の文字列 */
    fun removeExtension(pathOrUrl: String): String {
        val lastDotIndex = pathOrUrl.lastIndexOf('.')
        if (lastDotIndex == -1) return pathOrUrl
        return pathOrUrl.substring(0, lastDotIndex)
    }

    /** ディレクトリが存在しない場合は作成する。 @param targetFile 対象ディレクトリ @return 成功した場合は true */
    fun ensureDirectorySilently(targetFile: File): Boolean {
        if (targetFile.isFile) return false
        return if (targetFile.exists()) targetFile.canWrite()
        else targetFile.mkdirs()
    }

    /** 親ディレクトリが存在しない場合は作成する（エラー無視）。 @param targetFile 対象ファイル @return 成功した場合は true */
    fun ensureParentDirectorySilently(targetFile: File): Boolean {
        val parentFile = targetFile.parentFile
        if (parentFile == null) return false
        return ensureDirectorySilently(parentFile)
    }

    /**
     * ディレクトリが存在しない場合は作成する。エラー時は例外をスローする。
     * @param targetFile 対象ディレクトリ
     * @throws IOException 作成失敗または書き込み不可の場合
     */
    @JvmStatic fun ensureDirectory(targetFile: File) {
        if (targetFile.isFile) throw IOException("Target directory is a file")
        if (targetFile.exists()) {
            if (!targetFile.canWrite()) throw IOException("Target directory is not writable")
        } else if (!targetFile.mkdirs()) throw IOException("Unable to create target directory")
    }

    /**
     * 親ディレクトリが存在しない場合は作成する。
     * @param targetFile 対象ファイル
     * @throws IOException 親ディレクトリが存在しない、または作成失敗時
     */
    fun ensureParentDirectory(targetFile: File) {
        val parentFile = targetFile.parentFile
        if (parentFile == null) throw IOException("targetFile does not have a parent")
        ensureDirectory(parentFile)
    }
}
